apps/webapp/src/state/helpers.ts

```diff
@@ -1,18 +1,54 @@
 import { TransactionPlannerRequest } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1alpha1/view_pb';
 import { viewClient } from '../clients/grpc.ts';
 import { uint8ArrayToHex } from '@penumbra-zone/types';
+import { TransactionId } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/txhash/v1alpha1/txhash_pb';
+import { sha256Hash } from '@penumbra-zone/crypto-web';
 
-export const planWitnessBuildBroadcast = async (req: TransactionPlannerRequest) => {
-  const { plan } = await viewClient.transactionPlanner(req);
-  if (!plan) throw new Error('no plan in response');
+export const planWitnessBuildBroadcast = async (
+  req: TransactionPlannerRequest,
+  awaitDetection = true,
+) => {
+  const { plan: transactionPlan } = await viewClient.transactionPlanner(req);
 
-  const { transaction } = await viewClient.authorizeAndBuild({
-    transactionPlan: plan,
-  });
-  if (!transaction) throw new Error('no transaction in response');
+  if (!transactionPlan) throw new Error('no plan in response');
 
-  const { id } = await viewClient.broadcastTransaction({ transaction, awaitDetection: true });
-  if (!id) throw new Error('no id in broadcast response');
+  let transaction;
+  for await (const { status } of viewClient.authorizeAndBuild({ transactionPlan })) {
+    switch (status.case) {
+      case 'buildProgress':
+        console.log('buildProgress', status.value.progress);
+        break;
+      case 'complete':
+        transaction = status.value.transaction;
+        break;
+      default:
+        throw new Error(`unknown authorizeAndBuild status: ${status.case}`);
+    }
+  }
+  if (!transaction) throw new Error('did not build transaction');
 
-  return uint8ArrayToHex(id.inner);
+  const broadcastStream = viewClient.broadcastTransaction({ transaction, awaitDetection });
+
+  const expectId = new TransactionId({ inner: await sha256Hash(transaction.toBinary()) });
+
+  let detectionHeight;
```

Comment:
```suggestion
  let detectionHeight: bigint | undefined;
```

Reply:
linter and type system are already happy with it

Reply:
I meant more for human reasons than linting reasons: a type annotation here makes it immediately clear to readers what this variables can be.